src/main/java/com/milaboratory/mixcr/vdjaligners/VDJCAligner.java

```diff
@@ -171,12 +172,81 @@ public static VDJCAligner createAligner(VDJCAlignerParameters alignerParameters,
                 : new VDJCAlignerS(alignerParameters);
     }
 
-    public static Chains getPossibleDLoci(VDJCHit[] vHits, VDJCHit[] jHits) {
-        Chains chains = new Chains();
-        for (VDJCHit h : vHits)
-            chains = chains.merge(h.getGene().getChains());
-        for (VDJCHit h : jHits)
-            chains = chains.merge(h.getGene().getChains());
+    // public static Chains getPossibleDLoci(VDJCHit[] vHits, VDJCHit[] jHits) {
+    //     Chains chains = new Chains();
+    //     for (VDJCHit h : vHits)
+    //         chains = chains.merge(h.getGene().getChains());
+    //     for (VDJCHit h : jHits)
+    //         chains = chains.merge(h.getGene().getChains());
+    //     return chains;
+    // }
+
+    /*
+     *  Common utility methods
+     */
+
+    static Chains getVJCommonChains(final HasGene[] vHits, final HasGene[] jHits) {
+        return getChains(vHits).intersection(getChains(jHits));
+    }
+
+    /**
+     * Merge chains of all V/D/J/C hits.
+     *
+     * Allows null as input.
+     */
+    public static Chains getChains(final HasGene[] hits) {
+        if (hits == null || hits.length == 0)
+            return Chains.ALL;
+        Chains chains = hits[0].getGene().getChains();
+        for (int i = 1; i < hits.length; i++)
+            chains = chains.merge(hits[i].getGene().getChains());
         return chains;
     }
+
+    /**
+     * Calculates possible chains for D gene alignment, in the presence of following V, J and C genes.
+     *
+     * Allows nulls as input.
+     */
+    public static Chains getPossibleDLoci(HasGene[] vHits, HasGene[] jHits, HasGene[] cHits) {
+        Chains intersection = getChains(vHits)
+                .intersection(getChains(jHits))
+                .intersection(getChains(cHits));
+
+        if (!intersection.isEmpty())
+            return intersection;
+
+        // If intersection is empty, we are working with chimera
+        // lets calculate all possible D loci by merging
+        Chains chains = Chains.EMPTY;
+        if (vHits != null)
+            for (HasGene h : vHits)
```

Comment:
`chains = chains.merge(getChains(vHits))`  ?

Reply:
👌

Reply:
Actually, not exactly.

`null` check is still required, without it, `null` will lead to `ALL` return value in all cases. This is not the intended behaviour.

The same applies to zero-length arrays.

So, current code seems to be the most correct/clean.